### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by vitillo**
@mreid-moz r?

### CODE REVIEW

File: src/main/scala/utils/package.scala [hunk 1]
```diff
@@ -1,45 +1,42 @@
...
-                 "d2d_enabled" -> "D2DEnabled",
-                 "d_write_enabled" -> "DWriteEnabled",
-                 "vendor_id" -> "vendorID",
-                 "device_id" -> "deviceID",
-                 "subsys_id" -> "subsysID",
-                 "ram" -> "RAM",
-                 "gpu_active" -> "GPUActive",
-                 "first_load_uri" -> "firstLoadURI",
-                 "" -> "")
+  private val dict = Map("submission_url" -> "submissionURL",
```
Comment (mreid-moz) [inline src/main/scala/utils/package.scala]: While we're at it, can we give this variable a better name than `dict`?

File: src/main/scala/heka/Dataset.scala [hunk 2]
```diff
@@ -55,10 +60,25 @@ class Dataset private (bucket: String,
...
     val groups = ObjectSummary.groupBySize(summaries(fileLimit).toIterator)
     sc.parallelize(groups, groups.size).flatMap(x => x).flatMap(o => {
-      val hekaFile = s3Store.getKey(bucket, o.key)
-      for (message <- HekaFrame.parse(hekaFile, o.key)) yield message
+      try retry(3) {
+        val hekaFile = s3Store.getKey(bucket, o.key)
+        for (message <- HekaFrame.parse(hekaFile, o.key)) yield message
+      } catch {
+        case ex: Throwable =>
+          Logger.log.warn(s"Failure to read file ${o.key}: ${ex.getMessage}")
```
Comment (mreid-moz) [inline src/main/scala/heka/Dataset.scala]: For the first few attempts, we should make it an "info" level log message and say "attempt 1 of 3" etc, then if the final attempt fails, use "warn" or "error" log level, saying "all attempts failed".

**Comment by mreid-moz** [comment]
Can we add test coverage for the new retry logic?

File: src/main/scala/heka/Dataset.scala (hunk 2 above)
Comment (vitillo) [inline src/main/scala/heka/Dataset.scala]: Yes, I think this is not an issue as the final attempt throws an exception.

File: README.md [hunk 3]
```diff
@@ -33,13 +33,13 @@ See the [documentation for specific streams](https://github.com/mozilla/telemetr
 
 To generate a `DerivedStream`-based dataset `MyStream` for October 28, 2015 to October 29, 2015:
 ```bash
-sbt "run-main telemetry.DerivedStream --from-date 20151028 --to-date 20151029 MyStream"
+sbt "run-main com.mozilla.telemetry.DerivedStream --from-date 20151028 --to-date 20151029 MyStream"
```
Comment (mreid-moz) [inline README.md]: `org.mozilla.telemetry`?

File: src/main/scala/com/mozilla/telemetry/streams/E10sExperiment.scala [hunk 4]
```diff
@@ -119,7 +116,7 @@ case class E10sExperiment(experimentId: String, prefix: String) extends DerivedS
       .name("system").`type`().stringType().noDefault()
       .name("build").`type`().stringType().noDefault()
       .name("threadHangStats").`type`().stringType().noDefault()
-      .name("histograms").`type`().stringType().noDefault()
+      .name("org/mozilla/telemetry/histograms").`type`().stringType().noDefault()
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/streams/E10sExperiment.scala]: why did this field change names?

**Comment by mreid-moz** [comment]
We should still add test coverage for the retry behaviour in `Dataset`.

File: README.md (hunk 3 above)
Comment (vitillo) [inline README.md]: I have already used `com.mozilla.telemetry` for the hll package. As long as we own the domain it's fine.

**Comment by vitillo** [comment]
Note that I had to push the retry logic to `HekaFrame.parse` to avoid reading the entire file in `Dataset.records` which would increase memory pressure.

File: src/main/scala/com/mozilla/telemetry/SimpleDerivedStream.scala [hunk 5]
```diff
@@ -48,7 +44,7 @@ abstract class SimpleDerivedStream extends DerivedStream {
       hekaFile = bucket
       .getObject(key.key)
       .getOrElse(throw new Exception("File missing on S3"))
-      message <- HekaFrame.parse(hekaFile.getObjectContent(), hekaFile.getKey())
+      message <- HekaFrame.parse(hekaFile.getObjectContent())
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/SimpleDerivedStream.scala]: Don't need the parens after `getObjectContent`

File: src/main/scala/com/mozilla/telemetry/heka/Dataset.scala [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
 import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
 import org.json4s._
 import org.json4s.jackson.JsonMethods.parse
-import telemetry.ObjectSummary
-import telemetry.utils.{AbstractS3Store, S3Store}
+import com.mozilla.telemetry.utils.{AbstractS3Store, S3Store}
+import com.mozilla.telemetry.ObjectSummary
 import scala.io.Source
+import scala.util.{Failure, Success, Try}
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/heka/Dataset.scala]: These imports are no longer used

File: src/main/scala/com/mozilla/telemetry/heka/HekaFrame.scala [hunk 7]
```diff
@@ -63,20 +73,40 @@ object HekaFrame{
         Message.parseFrom(uncompressedMessage, 0, uncompressedLength)
       } catch {
         case ex: Throwable =>
-          Logger.log.warn(s"Failure to read compressed record of file $origin: ${ex.getMessage}")
           Message.parseFrom(messageBuffer)
       }
 
+      offset += 3 + headerLength + header.messageLength
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/heka/HekaFrame.scala]: Nit: add a comment that the `3` is one byte for the record separator, one for the header length, and one for the unit separator

File: src/main/scala/com/mozilla/telemetry/streams/E10sExperiment.scala [hunk 8]
```diff
@@ -52,7 +49,7 @@ case class E10sExperiment(experimentId: String, prefix: String) extends DerivedS
       .flatMap(x => x)
       .flatMap{ case obj =>
         val hekaFile = bucket.getObject(obj.key).getOrElse(throw new Exception("File missing on S3"))
-        for (message <- HekaFrame.parse(hekaFile.getObjectContent(), hekaFile.getKey()))  yield message }
+        for (message <- HekaFrame.parse(hekaFile.getObjectContent()))  yield message }
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/streams/E10sExperiment.scala]: nit: two spaces before `yield`

File: src/main/scala/com/mozilla/telemetry/streams/Longitudinal.scala [hunk 9]
```diff
@@ -98,7 +97,7 @@ case class Longitudinal() extends DerivedStream {
       .flatMap(x => x)
       .flatMap{ case obj =>
         val hekaFile = bucket.getObject(obj.key).getOrElse(throw new Exception("File missing on S3: %s".format(obj.key)))
-        for (message <- HekaFrame.parse(hekaFile.getObjectContent(), hekaFile.getKey()))  yield message }
+        for (message <- HekaFrame.parse(hekaFile.getObjectContent()))  yield message }
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/streams/Longitudinal.scala]: nit: too many spaces before `yield`

File: src/main/scala/com/mozilla/telemetry/utils/package.scala [hunk 10]
```diff
@@ -1,45 +1,42 @@
...
+    val matcher = pattern.matcher(name)
     val output = new StringBuilder
 
     while (matcher.find()) {
-      if (output.length > 0)
-        output.append("_");
+      if (output.length > 0) {
+        output.append("_")
+      }
       output.append(matcher.group().toLowerCase);
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/utils/package.scala]: nit: remove trailing `;`

File: src/main/scala/com/mozilla/telemetry/utils/package.scala [hunk 11]
```diff
@@ -1,45 +1,42 @@
...
+    "speed_mhz" -> "speedMHz",
+    "d2d_enabled" -> "D2DEnabled",
+    "d_write_enabled" -> "DWriteEnabled",
+    "vendor_id" -> "vendorID",
+    "device_id" -> "deviceID",
+    "subsys_id" -> "subsysID",
+    "ram" -> "RAM",
+    "gpu_active" -> "GPUActive",
+    "first_load_uri" -> "firstLoadURI",
+    "" -> "")
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/utils/package.scala]: why is the empty string in here?

File: src/main/scala/com/mozilla/telemetry/utils/package.scala [hunk 12]
```diff
@@ -1,45 +1,42 @@
...
-                 "d2d_enabled" -> "D2DEnabled",
-                 "d_write_enabled" -> "DWriteEnabled",
-                 "vendor_id" -> "vendorID",
-                 "device_id" -> "deviceID",
-                 "subsys_id" -> "subsysID",
-                 "ram" -> "RAM",
-                 "gpu_active" -> "GPUActive",
-                 "first_load_uri" -> "firstLoadURI",
-                 "" -> "")
+  private val dict = Map("submission_url" -> "submissionURL",
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/utils/package.scala]: rename `dict` something more descriptive, such as `specialCases`

File: src/test/scala/com/mozilla/telemetry/DatasetTest.scala [hunk 13]
```diff
@@ -74,7 +85,7 @@ class DatasetTest extends FlatSpec with Matchers{
       case "main" => true
     }.summaries(Some(1)).toList
 
-    assert(files == List(ObjectSummary("x", 1)))
+    files should be (List(ObjectSummary("x", 1)))
   }
 
   "Records" can "be fetched" in {
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/DatasetTest.scala]: On following line, s/KPI/DatasetTest/

File: src/test/scala/com/mozilla/telemetry/DatasetTest.scala [hunk 14]
```diff
@@ -90,7 +101,26 @@ class DatasetTest extends FlatSpec with Matchers{
...
 
-      assert(records.count() == 42)
+      records.count() should be (42)
+    } finally {
+      sc.stop()
+    }
+  }
+
+  "Reads from S3" should "be retried in case of failure" in {
+    val sparkConf = new SparkConf().setAppName("KPI")
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/DatasetTest.scala]: s/KPI/DatasetTest/

**Comment by mreid-moz** [comment]
Looks good, r+

**Comment by vitillo** [comment]
@mreid-moz @maurodoglio please update your scheduled Scala jobs as the package structure has changed

**Pull request merged**